samples/subsys/settings/src/main.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * NSETTINGS Sample for Zephyr
+ *
+ * Copyright (c) 2018 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <zephyr.h>
+#include <misc/reboot.h>
+#include <device.h>
+#include <string.h>
+#include <settings/settings.h>
+
+/* 1000 msec = 1 sec */
+#define SLEEP_TIME      2000
+
+u8_t reset_counter;
+
+static int ps_set(const char *key, size_t len, settings_read_cb read,
+		  void *cb_arg)
+{
+	const char *dyn_name;
+	char argv[32];
+
+	if (key == NULL) {
+		return 0;
+	}
+
+	/* use settings_name_cmp to compare with a fixed key */
+	if (settings_name_cmp(key, "ra0", NULL)) {
```

Comment:
It would also match "ra0/foo/boo"

Reply:
Yes it would, to be sure there is nothing left the next element should be used and compared with NULL. This is only a sample, it is not my intention to include this in the final PR.

Reply:
In might be worth to return different values by settings_name_cmp, for example:
1 - full match
0 - partial match
-EINVAL - no match.

Then we can use ```if (0 < settings_name_cmp(...)``` if we wish full match only, ```if (0 <= settings_name_cmp(...)``` if we wish to accept also partial match. It is just one line change and simplifies the settings_name_cmp usage (no need additional ```next``` checking to make sure if we have a full match.

Reply:
I originally had it like that, but imho it's easier to understand if it only returns 0/1. Otherwise there are comparisons: `== 0, >= 0, > 0` ... 

`settings_name_cmp()` and `settings_name_split()` should in modules also rarely be called with `next=NULL` and should be discouraged.

Reply:
If you are using something with cmp as suffix Id expect it to return something like strcmp, if that is not the case then please use another suffix or change the behavior to match.

Reply:
To explain such "complications" the function documentation should be used.
I still think that this:

```
if (0 < settings_name_cmp(key, "ra0", NULL) {
}
```

is much easier to read than:

```
const char *next;

if (settings_name_cmp(key, "ra0", &next) && !next) {
}
```

If you wish to continue this that way - please fix the current usage of the function to cover the described problem (processing the values that are not fully matched).

Reply:
@Vudentz 
Maybe right here
> change the behavior to match.

Match seems nice: ```settings_name_match``` ;)

Reply:
ok, will need to change the name to `settings_name_steq()`, short for start equal.

Reply:
> To explain such "complications" the function documentation should be used.
> I still think that this:
> 
> ```
> if (0 < settings_name_cmp(key, "ra0", NULL) {
> }
> ```
> 
> is much easier to read than:
> 
> ```
> const char *next;
> 
> if (settings_name_cmp(key, "ra0", &next) && !next) {
> }
> ```
> 
> If you wish to continue this that way - please fix the current usage of the function to cover the described problem (processing the values that are not fully matched).

In this case it is easier, however if we look at a real case it would be something like:
```
const char *next;

if (settings_name_cmp(key, "ra0", &next) {
   if ((!next) && (len == expected_size)) {
      /* read it */
   }
}
```
if we encourage this use method, accidents should be rare.

Reply:
Sadly it would not look so simple:

```
const char *next;

if (settings_name_steq(key, "ra0", &next) && !next) {
   if (len == expected_size) {
      /* read it */
      return 0;
   } else {
      return len;
   }
}

return -NOTSUP;
```

Reply:
I am not sure if the last return is required. It is just some data that is there that is not processed by the set routine, this should be ignored, not trigger an error. IMO settings should not say I have found some data that should not be there, it should ignore it.

Reply:
The function above should decide if it ignores it and this is how it is done when using data loading from FLASH.
But for CLI injection, what if the user misspells the name - should it silently accept it or should it rather display an error?

Reply:
You are right, I forgot about the CLI injection (runtime_set()). I even never realized this return value is only used by the CLI injection. Thanks for pointing this out.